### DISCUSSION TITLE
Custom package name registry url splitter for go

### DISCUSSION THREAD

**Discussion by ManuelB** [renovatebot/renovate#26408] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

latest

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I have not seen this working

### Describe the problem

I implemented a config customPackageNameRegistryUrlSplitter this makes it possible to lookup dependencies in custom gitlab repositories where a part of the url is part of the registry url and another part is the package name e.g.:

...
    {
      "matchHost": "example.com/gitlab",
      "customPackageNameRegistryUrlSplitter": "(example.com/gitlab/)(.*).git",
      "token": process.env.RENOVATE_TOKEN
    },
...

Pull Request is here: https://github.com/renovatebot/renovate/pull/26399

This bug only appears on private gitlab instances when using go. Unfortunately I can't give you access to the gitlab instance and I am.not able to set up a whole git lab for reproducing this error.

### Relevant debug logs

<details><summary>Logs</summary>

```
...
Failed to look up go package example.com/gitlab/services/project/libraries/common.git
...
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
You have described your solution, in the part where you're asked to describe the problem. The solution wasn't ideal, so I'd like to learn more about the problem

**Comment by ManuelB** [comment]
@rarkins the problem is that we have a go repository with a go.mod file. This go mod file contains modules that point to a private gitlab repository e.g.:

```
example.com/gitlab/services/project/libraries/common.git
```

When the normal gitlab-tags method is used to fetch the releases the system will do the following split:

registryUrl: example.com
packageName: gitlab/services/project/libraries/common

When this package name is used to do the lookup for releases in the gitlab API it ends with a 404 error:

https://example.com/gitlab/api/v4/projects/gitlab%F6services%F6project%F6libraries%F6common/repository/tags

I hope that this illustrates the problem. The log will show:

...
Failed to look up go package example.com/gitlab/services/project/libraries/common.git
...

**Reply by rarkins** [reply]
Is this the same gitlab server which Renovate itself is running against? i.e. it's running one repo A, looking up a Go library on repo B, both on the same gitlab server?

If so then what is your `endpoint` in Renovate set to (i.e. `RENOVATE_ENDPOINT`)

**Reply by ManuelB** [reply]
Yes it is the same server. In the config the endpoint is set like this:

```
  "platform": "gitlab",
  "endpoint": "https://example.com/gitlab/api/v4/",
```

**Reply by rarkins** [reply]
You don't have a `https://` prefix for your endpoint?

**Reply by ManuelB** [reply]
Not in the go package name. This is normal for go packages.

**Reply by ManuelB** [reply]
Here is a typical example of a go.mod file:

https://gitlab.com/gitlab-org/gitaly/-/blob/master/go.mod

**Reply by rarkins** [reply]
I asked about endpoint, not the go package

**Reply by ManuelB** [reply]
@rarkins the endpoint has https:// in it. It is: https://example.com/gitlab/api/v4/. Can you explain more what information you required?

A packageRule like:

```
    {
      "matchDatasources": ["gomod"],
      "registryUrls": ["https://example.com/gitlab"]
    }
```

did not solve the problem

**Reply by rarkins** [reply]
@ManuelB please edit/correct your earlier comment above where you showed an invalid/incorrect `endpoint` value

**Reply by ManuelB** [reply]
@rarkins I updated the comment

**Reply by rarkins** [reply]
I've added backticks so that github doesn't drop the https:// prefix when displaying

**Reply by rarkins** [reply]
What the above means is that Renovate has enough information already to determine the correct registryUrls/packageName split for this go dependency and that explicit configuration shouldn't be necessary.

I believe here is the logic which is currently splitting the URL incorrectly: https://github.com/renovatebot/renovate/blob/5d95ece27110bf1208feb3f2e30dab692d10471f/lib/modules/datasource/go/base.ts#L120-L150

Somewhere in there we could add logic to compare the goSourceUrl to the `endpoint` value and deduce the correct split. The `endpoint` value should be available available via `GlobalConfig.get('endpoint')`

**Reply by ManuelB** [reply]
@rarkins in my case this is true but this might not be true for other cases. I put a lot of effort in the first solution and I like the solution. Nevertheless if you want me to implement a less powerful solution and this is the only chance to get this pull request accepted I will implement it.

**Reply by ManuelB** [reply]
@rarkins endpoint in my case will contain: https://example.com/gitlab/api/v4/ for this URL I don't know how to split it. So the question is what is the registry url:

https://example.com/gitlab/api/v4/?
https://example.com/gitlab/api/?
https://example.com/gitlab/?

Others might have other urls like:
https://gitlab.example.com/
https://example.com/gitlab/latests/api/v4

**Reply by rarkins** [reply]
You can assume any api/v4 suffix can be removed

**Reply by ManuelB** [reply]
@rarkins I will implement your suggested solution
